Add UI for unregistered block types

brandonpayton opened the pull request:
## Description
This PR adds UX for unregistered block types found in the post content.

A notice is shown in the place of each unregistered block type, offering the ability to remove the block or leave its content intact. If the core Custom HTML block is present and the unregistered block has HTML content, the notice also offers the opportunity to keep the block HTML as a Custom HTML block.

⚠️ Child blocks are not preserved. This is consistent with today's fallback behavior. Parsed blocks do not yet contain information about where their child blocks were found, so it is not possible to reconstitute child blocks within an unregistered parent. This is important to fix but should probably be done after a separate PR that updates the parser to provide such information. See #8214 for what appears to be a related issue.   

Contributes to #7811.

## How has this been tested?

Unit tests. It has also been manually tested by:

1. Creating custom block types.
2. Removing those types and reloading to observe missing block warnings.
3. Changing other content, saving, re-enabling the custom block types, and reloading to observe the original block content was preserved and loaded correctly again.

## Screenshots <!-- if applicable -->
<img width="969" alt="screen shot 2018-09-06 at 10 26 43 am" src="https://user-images.githubusercontent.com/530877/45191941-aac86400-b1fa-11e8-9716-338872c34e82.png">

## Types of changes

Today, we have an "unknown fallback" that handles both non-block content and unregistered block types. This PR splits the fallback in two, a freeform content fallback and an unregistered type fallback, and deprecates the "unknown fallback".

In discussion with @karmatosed, we want the unregistered block fallback to appear more like a warning than a block, so we hide its block outline and override its block sidebar to appear as if no block is selected.

brandonpayton committed:
Add block to represent unregistered block types

karmatosed commented:
Design wise could we go more with this please?

![42715496-372cdf0c-86ef-11e8-8692-3e398df4a18e](https://user-images.githubusercontent.com/253067/43418184-16076ad8-9435-11e8-92e1-f43dc353e9c5.png)

This was outlined in comments and be great to have consistency. Ideally greying out the background - is that possible?

karmatosed reviewed (changes requested):
Design needs to if possible reflect the proposed design. This seems to add a different style.

brandonpayton commented:
Thanks for taking a look, @karmatosed.

> Design needs to if possible reflect the proposed design. This seems to add a different style.

Yeah, I used the same `Warning` component we use for block validation errors and noticed it didn't match your design, probably due to multiple lines. It was one of things I needed to follow up on. (The point of mentioning this is to let you know I saw and had regard for your designs but hadn't gotten to fixing this as it was a WIP).

In a [previous comment](https://github.com/WordPress/gutenberg/issues/7811#issuecomment-404963199), you mentioned:
> I will also note this new notice style has a bit of a problem of scaling to multiple lines but that would need to be tackled in the other issue.

Was this referring to our other notices like this one, and if so, would you be up for pointing me to the other issue that deals with the notice style so I can be sensitive to what we're doing there?

Regarding the design you shared above (and in the originating issue):

1. Do I understand correctly that our message lines should wrap to the left of the button while the button remains in the upper right?
2. Which background should be greyed out? I see there is a bit of grey background on the left hand side of your image, but I don't think I'm fully understanding your intent.

brandonpayton committed:
Make this feel more like a message than a block

brandonpayton committed:
Add a Remove Block button

brandonpayton committed:
Try to clarify block sidebar

brandonpayton committed:
Remove unused supports flag

brandonpayton committed:
Fix removal for content-free, unregistered block

brandonpayton committed:
Avoid showing block settings for unregistered blocks

brandonpayton commented:
There is a CSS outline transition bug that needs fixed and a related `!important` (added during troubleshooting) that will be removed.

brandonpayton committed:
Stop outline transition flicker

brandonpayton commented:
The outline transition is now fixed.

brandonpayton committed:
Switch back to removing via block menu

brandonpayton committed:
Disallow making missing blocks reusable

brandonpayton committed:
Update deprecation target

brandonpayton committed:
Improve naming of freeform fallback

brandonpayton committed:
Additional cleanup

brandonpayton commented:
@karmatosed and I met out of band and discussed her latest feedback. Here's my understanding of the outcome:

* We should should use the block menu for removing an unregistered block rather than providing a "Remove block" button. The addition of the button was due to a misunderstanding from an earlier conversation.
* The bottom location of the "Keep as HTML" button is OK because it how our Warning component currently behaves for longer text.
* We won't try to display the warning as a content overlay. The reason is that styles are missing for unregistered blocks, making it unlikely we'll render them well.

brandonpayton commented:
Hi Karmatosed, I've rebased and updated the UI based on our discussion. Here's how the latest looks:
<img width="969" alt="screen shot 2018-09-06 at 10 26 43 am" src="https://user-images.githubusercontent.com/530877/45191941-aac86400-b1fa-11e8-9716-338872c34e82.png">

What do you think?

Recently, the location of the block menu changed to the block toolbar, and that looks a bit odd since we hid the block outline here. I'm wondering whether this is still OK to merge, and we can adjust that in a follow up PR.

mtias commented:
There has to be an option about not doing anything. If we can map a block with a plugin, we can offer to activate it it.

mtias commented:
Also, if the block has HTML content we should somehow show it as it'd help the user determine what to do — should be similar to the invalid block preview.

brandonpayton commented:
As a follow-up to my last comment, I see the preview you're talking about with the invalid block warning. It looks like I was missing it because the preview had a small height and was hidden behind the warning overlay in my tests. My concern about showing a preview is that we don't have styles for missing blocks, so the preview will just have default browser styles. Is that fine with you?

talldan commented:
@brandonpayton I think it might have been this that's the preview being referred to:
https://github.com/WordPress/gutenberg/pull/7995

brandonpayton commented:
Thanks, @talldan! That's interesting. I didn't know about that option.

mcsf commented:
> My concern about showing a preview is that we don't have styles for missing blocks, so the preview will just have default browser styles. Is that fine with you?

A valid concern, but there is still value in rendering a preview in that this can let the user know how the block still looks (and a plainly styled block may continue looking just the same), assuage their fear of content loss, and — if the front-end preview looks acceptable to them — let them choose to keep the block as is.

mcsf commented on packages/blocks/src/api/registration.js:
```diff
@@ -191,6 +191,11 @@ export function unregisterBlockType( name ) {
 export function setUnknownTypeHandlerName( name ) {
+	deprecated( 'setUnknownTypeHandlerName', {
+		plugin: 'Gutenberg',
+		version: '4.0',
```
These versions need to be bumped by at least .2 :)

mcsf replied:
Both in this file and in `store/*`, can't we simply treat the deprecated `*UnknownType*` functions as aliases for their `*FreeformContent*` counterparts? If I read the diff correctly, one deprecated function keeps calling a deprecated action creator, which keeps outputting an action of deprecated type.

brandonpayton replied:
Following up on this one [here](https://github.com/WordPress/gutenberg/pull/8274#discussion_r223131741).

brandonpayton replied:
> These versions need to be bumped by at least .2 :)

I ended up bumping them to 4.4 because we already have documented deprecations for that version, and this is quite a last minute API change.

> If I read the diff correctly, one deprecated function keeps calling a deprecated action creator, which keeps outputting an action of deprecated type.

Based on your suggestion, the deprecated functions now call their `*FreeformContent*` counterparts. This is simpler, and calling a deprecated block registration function no longer leads to multiple deprecation warnings.

mcsf reviewed (commented):
Product-wise, this looks good so far! It now needs the previewing ability.

In terms of stability, we need some decision on whether to change the API, and, if so, on the flow of the deprecation code, which to me seemed harder to read.

brandonpayton commented on docs/reference/deprecated.md:
```diff
@@ -6,6 +6,8 @@ Gutenberg's deprecation policy is intended to support backwards-compatibility fo
 - `wp.i18n.getI18n` has been removed. Use `__`, `_x`, `_n`, or `_nx` instead.
 - `wp.i18n.dcnpgettext` has been removed. Use `__`, `_x`, `_n`, or `_nx` instead.
+- `wp.blocks.setUnknownTypeHandlerName` has been removed. Please use `setFreeformContentHandlerName` and `setUnregisteredTypeHandlerName` instead.
+- `wp.blocks.getUnknownTypeHandlerName` has been removed. Please use `getFreeformContentHandlerName` and `getUnregisteredTypeHandlerName` instead.
```
From [another comment]():
> Both in this file and in `store/*`, can't we simply treat the deprecated `*UnknownType*` functions as aliases for their `*FreeformContent*` counterparts? If I read the diff correctly, one deprecated function keeps calling a deprecated action creator, which keeps outputting an action of deprecated type.

It seems like it might be better to just stick with `setUnknownTypeHandlerName` since the function isn't often used. An "unknown" type might be distinct enough from an "unregistered" type to leave as-is, but I am happy to use an alias if you prefer.

brandonpayton replied:
I've been away from this PR for too long. I am reminded now that the naming for the store is less clear with `setFallbackBlockName`. Aliasing seems to make more sense because of this case.

brandonpayton committed:
Add missing reducer unit tests

brandonpayton committed:
Remove cruft left by missing block iteration

brandonpayton committed:
Clean up deprecation docs and control flow

brandonpayton committed:
Clarify missing block case in block inspector

brandonpayton committed:
Stop using deprecated getFallbackBlockName

brandonpayton committed:
Treat unregistered block case as a warning

brandonpayton committed:
Update unit tests

brandonpayton commented on packages/block-library/src/missing/index.js:
```diff
@@ -0,0 +1,95 @@
+	withSelect( ( select, { clientId } ) => {
+		const { getBlock } = select( 'core/editor' );
+		return {
+			block: getBlock( clientId ),
```
Fixed. Thank you.

brandonpayton commented on packages/editor/src/components/block-inspector/index.js:
```diff
@@ -24,7 +24,12 @@ const BlockInspector = ( { selectedBlock, blockType, count } ) => {
 
-	if ( ! selectedBlock ) {
+	/*
+	 * In addition to the actual case of no-block-selected,
```
Agreed. That's much better. I made the change.

brandonpayton commented:
Hi @mcsf, I made the following updates based on your review:

* Added a preview to the warning.
*  Updated the deprecated functions to be aliases of the freeform content handler functions. The deprecated block registration functions no longer call deprecated block store functions that would generate a second warning.
* Simplified and clarified as suggested.

I also updated the docs, adjusted some newer things that were calling the deprecated getter methods, and added some missing reducer unit tests.

Here's a shot of the latest:
<img width="995" alt="screen shot 2018-10-08 at 10 57 07 pm" src="https://user-images.githubusercontent.com/530877/46653432-e2f4f680-cb5a-11e8-9b08-b8875a2443da.png">

Some things that could be done in a follow-up PR:

* Rename `block-unknown-convert-button` to `block-freeform-convert-button` because it only works for the freeform block. If we enable it for unregistered blocks, its implementation parses the HTML and generates an identical unregistered block, making it appear as if nothing happened. This is because unregistered block serializationn preserves the raw, delimited content of the original block.
* Create a `block-unregistered-convert-button` if we want to support conversion of unregistered blocks. The implementation would likely be similar to `block-unknown-convert-button`, but the HTML content would be taken from the block's `originalUndelimitedContent` attribute.
* Preserve inner block content. It is my understanding that our parser does not yet provide the indices of inner blocks within their parent, making it impossible to reconstitute the original content without a registered block implementation. cc @dmsnell in case this is now fixed.

brandonpayton commented:
I will be AFK for the rest of the week, and @talldan has agreed to keep an eye on this in the meantime. Thanks, Dan!

karmatosed commented:
Design wise this seems on track now so removing the feedback label and passing for that review.

mcsf committed:
Lower deprecation version to 4.2

mcsf commented:
Thanks for the work, @brandonpayton. So far this looks pretty good.

- I've rebased against `master`.
- I've lowered the deprecation version to 4.2 (bf89150), given our core merge timeline.

mcsf commented:
Noting a bug that is not introduced by this PR, merely highlighted by it:

![html-styles](https://user-images.githubusercontent.com/150562/46867111-c804d980-ce1b-11e8-8a8f-36409709211b.gif)

It's an issue in the HTML block and other users of `iframe`, in which editor styles aren't respected. Tracked in #10547. It doesn't block this PR.

mcsf committed:
chore: Build docs

mcsf committed:
Minor refactor (isFallbackBlock); wrap lines

mcsf commented on packages/blocks/src/api/parser.js:
```diff
@@ -394,54 +398,61 @@ export function getMigratedBlock( block ) {
 		innerHTML,
 	} = blockNode;
+	const freeformContentFallbackBlock = getFreeformContentHandlerName();
+	const unregisteredFallbackBlock = getUnregisteredTypeHandlerName() || freeformContentFallbackBlock;
```
This initially struck me as odd, but upon testing the actual behavior, it's not bad to default to Freeform if Unregistered is not present. 👍 (Although, who'd have one and not the other?)

aduth commented on packages/block-library/src/missing/index.js:
```diff
@@ -0,0 +1,91 @@
+	return (
+		<Fragment>
+			<Warning actions={ actions }>
+				<span dangerouslySetInnerHTML={ { __html: messageHTML } } />
```
In this case, dangerous is _actually dangerous_. Localized strings should be considered equivalent to user input for sanitization purposes. 

> since translations should not be considered trusted strings, be sure to sanitize the result before echoing.

https://codex.wordpress.org/I18n_for_WordPress_Developers#HTML

Related: #9846

mcsf replied:
-> #10626

brandonpayton replied:
Thanks very much for catching this, @aduth, and for alerting me to the need for care here.

@mcsf, thank you for the fix. <3

aduth replied:
>FWIW this isn't entirely accurate. WordPress.org treats stranslations as trusted strings.

Is the Codex wrong then? Or is the article not applicable for core development?

swissspidy replied:
Both I guess :-)

aduth commented on packages/blocks/src/api/parser.js:
```diff
@@ -394,54 +398,61 @@ export function getMigratedBlock( block ) {
 
 	// Trim content to avoid creation of intermediary freeform segments.
-	innerHTML = innerHTML.trim();
+	const originalUndelimitedContent = innerHTML = innerHTML.trim();
```
This is [too clever](https://softwareengineering.stackexchange.com/questions/25276/why-is-cleverness-considered-harmful-in-programming-by-some-people). I'm still not totally sure I understand why / what we're doing with this line.

brandonpayton replied:
Thanks, @aduth. The intent wasn't cleverness, but I see what you mean. We want `innerHTML` to be trimmed, and saving its value at that point is a separate concern.

The missing block needs the original undelimited content for determining whether the unregistered block has actual HTML content, for rendering a preview of the block HTML, and for converting to an HTML block. `originalUndelimitedContent` exists because [this pre-existing line](https://github.com/WordPress/gutenberg/pull/8274/files/79222f46ab26c11f4f6b8e8289b4ad811a2faf76#diff-7dae7b3890cf27659680c926b6992e31R438) changes the value of `innerHTML` to be the delimited HTML content. The delimited content is used in saving the value of the block.

The above approach, which I am responsible for, is a bit weird IMO. It would be more straightforward to:
1. Get rid of `originalUndelimitedContent`.
1. Stop overwriting `innerHTML` with the delimited content.
1. Add an `originalAttributes` attribute to `core/missing`.
1. Update `core/missing` to call `getCommentDelimitedContent` with `originalName`, `originalAttributes` and `originalContent` to create the save content.

If you agree this is cleaner, I can create a follow-up PR to make the change.

aduth replied:
I actually don't have much of a problem with `originalUndelimitedContent`, I just think we could be clearer / more verbose with how we assign / document its intent.

Also, I think with this type of assignment we're inadvertently assigning `innerHTML` as global outside the scope of the function:

![image](https://user-images.githubusercontent.com/1779930/47032632-d49d7080-d140-11e8-8b87-fffd553e7c7c.png)

brandonpayton replied:
> I actually don't have much of a problem with originalUndelimitedContent, I just think we could be clearer / more verbose with how we assign it / document its intent.

✅ thanks
I don't love `originalUndelimitedContent` but plan to leave it after documenting its purpose and separating the assignment.

I don't believe this assignment should be creating a global since `innerHTML` is declared in the function scope [here](https://github.com/WordPress/gutenberg/pull/8274/files/79222f46ab26c11f4f6b8e8289b4ad811a2faf76#diff-7dae7b3890cf27659680c926b6992e31R405).

aduth commented on packages/block-library/src/missing/index.js:
```diff
@@ -0,0 +1,91 @@
+			<Warning actions={ actions }>
+				<span dangerouslySetInnerHTML={ { __html: messageHTML } } />
+			</Warning>
+			<div>
```
For what reason do we need the `div`?

mcsf commented:
One thing I'm noticing in master, @brandonpayton, is how Unregistered blocks are not selectable:

- If I focus one such block, the Block sidebar reads _No block selected._
- If I try to multi-select with the keyboard, I can't start the process on such a block. Arrow-based selection seems fine, even though the visual cues for selected state could be clearer.

brandonpayton commented:
I created #10712 to track the selection highlight bug.

gAllegr commented:
> Here's a screen cap of my experience:
> ![selecting-block-warnings](https://user-images.githubusercontent.com/530877/47121852-320cec80-d229-11e8-8f8d-951fbc66dfda.gif)

Speaking as a translator, there are two strings with _Resolve_ word ([String 1](https://translate.wordpress.org/projects/wp-plugins/gutenberg/dev/it/default?filters%5Boriginal_id%5D=6745337) and [String 2](https://translate.wordpress.org/projects/wp-plugins/gutenberg/dev/it/default?filters%5Boriginal_id%5D=6745336) ) which are not so self-explicative.

It is possible to add a comment to these strings?

mcsf commented:
> This is with Chrome. In what browser are you seeing this issue?

It would've either been Firefox Developer Edition or Safari, my more commonly used browsers.
